### PR TITLE
[module] add mutation support for forward_pre_hook and forward_hook

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#22285 [module] add mutation support for forward_pre_hook and forward_hook**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D16022491/)

Previously forward hooks are expected to return None, this PR adds the support to overwrite input and output in `forward_pre_hook` and `forward_hook`, this is used to implement inserting quant/dequant function calls around forward functions.

Differential Revision: [D16022491](https://our.internmc.facebook.com/intern/diff/D16022491/)

### CODE REVIEW

File: test/test_nn.py [hunk 1]
```diff
@@ -766,6 +748,22 @@ def bw_hook(module, grad_input, grad_output):
...
+            return (torch.nn.functional.relu(input[0]),)
+
+        def forward_hook(m, input, output):
+            return -output
+
+        module.register_forward_pre_hook(forward_pre_hook)
+        module.register_forward_hook(forward_hook)
+        output = module(input)
+        expected_res = -torch.nn.functional.linear(torch.nn.functional.relu(input), module.weight, module.bias)
+        self.assertEqual(output, expected_res)
```
Comment (raghuramank100) [inline test/test_nn.py]: Does backward get computed correctly now? This is important for QAT.  Previously it would not have mattered as hooks did not modify input or output. Can you add a test for backward correctness?

**Review by raghuramank100 — CHANGES_REQUESTED** [review]
Please check if behavior for backward is correct.

File: torch/nn/modules/module.py [hunk 2]
```diff
@@ -524,17 +524,17 @@ def _slow_forward(self, *input, **kwargs):
 
     def __call__(self, *input, **kwargs):
         for hook in self._forward_pre_hooks.values():
-            hook(self, input)
+            result = hook(self, input)
+            if result:
```
Comment (apaszke) [inline torch/nn/modules/module.py]: This seems wrong? We either need to clearly document that you're supposed to return a tuple from this, or we have to auto-wrap the input in a tuple if a single value has been returned from the hook. Think about the case when someone returns a single tensor.

**Review by apaszke — CHANGES_REQUESTED** [review]

File: test/test_nn.py (hunk 1 above)
Comment (apaszke) [reply test/test_nn.py]: Hooks are no different than any other Python code, so they don't really affect how backward is computed. We don't need tests for that.

Comment (jerryzh168) [reply test/test_nn.py]: Yeah we just record the backward on Tensors, so I think it will compute backward correctly as long as the fakequant gradient is correct.

File: torch/nn/modules/module.py (hunk 2 above)
Comment (jerryzh168) [reply torch/nn/modules/module.py]: sure, I'll wrap the input

Comment (jerryzh168) [reply torch/nn/modules/module.py]: will there be a case when the forward function accepts a tuple as input?

Comment (apaszke) [reply torch/nn/modules/module.py]: Yes, this is very much possible. If we want to go down the wrapping path, then I'd create a singleton tuple if `not isinstance(result, tuple)`.

File: torch/nn/modules/module.py [hunk 3]
```diff
@@ -524,17 +528,20 @@ def _slow_forward(self, *input, **kwargs):
 
     def __call__(self, *input, **kwargs):
         for hook in self._forward_pre_hooks.values():
-            hook(self, input)
+            result = hook(self, input)
+            if isinstance(result, tuple):
+                input = result
+            else:
+                if result is not None:
+                    input = (result,)
```
Comment (apaszke) [inline torch/nn/modules/module.py]: nit: I think this is slightly more readable:
```python
if result is not None:
    if not isinstance(result, tuple):
        result = (result,)
    input = result
```

File: torch/nn/modules/module.py [hunk 4]
```diff
@@ -462,9 +462,11 @@ def register_forward_pre_hook(self, hook):
         The hook will be called every time before :func:`forward` is invoked.
         It should have the following signature::
 
-            hook(module, input) -> None
+            hook(module, input) -> None or modified input
 
-        The hook should not modify the input.
+        The hook can modify the input. User can either return a tuple or a
+        single modified value in the hook. We will wrap the value into a tuple
+        if a single value is returned.
```
Comment (apaszke) [inline torch/nn/modules/module.py]: ... if a single value is returned _(unless that value is already a tuple)_.

**Review by apaszke — APPROVED** [review]
Looks good. Minor nits.

File: torch/nn/modules/module.py (hunk 2 above)
Comment (jerryzh168) [reply torch/nn/modules/module.py]: So if the forward function accepts a tuple as input, should we wrap the tuple again in another singleton tuple?

Comment (apaszke) [reply torch/nn/modules/module.py]: Well what we're doing is a bit magical and in general hard to do, but I think it's also pretty consistent and user-friendly. We can document the edge cases, but I don't think they will be very common.
